sentry_sdk/integrations/rust_tracing.py

```diff
@@ -0,0 +1,259 @@
+
+
+class RustTracingLayer:
+    def __init__(
+        self,
+        origin: str,
+        event_type_mapping: Callable[
+            [Dict[str, Any]], EventTypeMapping
+        ] = default_event_type_mapping,
+        span_filter: Callable[[Dict[str, Any]], bool] = default_span_filter,
+    ):
+        self.origin = origin
+        self.event_type_mapping = event_type_mapping
+        self.span_filter = span_filter
+
+    def on_event(self, event: str, _span_state: TraceState) -> None:
+        deserialized_event = json.loads(event)
+        metadata = deserialized_event.get("metadata", {})
+
+        event_type = self.event_type_mapping(metadata)
+        if event_type == EventTypeMapping.Ignore:
+            return
+        elif event_type == EventTypeMapping.Exc:
+            process_exception(deserialized_event)
+        elif event_type == EventTypeMapping.Breadcrumb:
+            process_breadcrumb(deserialized_event)
+        elif event_type == EventTypeMapping.Event:
+            process_event(deserialized_event)
+
+    def on_new_span(self, attrs: str, span_id: str) -> TraceState:
+        attrs = json.loads(attrs)
+        metadata = attrs.get("metadata", {})
+
+        if not self.span_filter(metadata):
+            return None
+
+        module_path = metadata.get("module_path")
+        name = metadata.get("name")
+        message = attrs.get("message")
+
+        if message is not None:
+            sentry_span_name = message
+        elif module_path is not None and name is not None:
+            sentry_span_name = f"{module_path}::{name}"  # noqa: E231
+        elif name is not None:
+            sentry_span_name = name
+        else:
+            sentry_span_name = "<unknown>"
+
+        kwargs = {
+            "op": "native_extension",
+            "name": sentry_span_name,
+            "origin": self.origin,
+        }
+
+        scope = sentry_sdk.get_current_scope()
+        parent_sentry_span = scope.span
+        if parent_sentry_span:
+            sentry_span = parent_sentry_span.start_child(**kwargs)
+        else:
+            sentry_span = scope.start_span(**kwargs)
+
+        fields = metadata.get("fields", [])
+        for field in fields:
+            sentry_span.set_data(field, attrs.get(field))
+
+        scope.span = sentry_span
+        return (parent_sentry_span, sentry_span)
+
+    def on_close(self, span_id: str, span_state: TraceState) -> None:
+        if span_state is None:
+            return
+
+        parent_sentry_span, sentry_span = span_state
+        sentry_span.finish()
+        sentry_sdk.get_current_scope().span = parent_sentry_span
+
+    def on_record(self, span_id: str, values: str, span_state: TraceState) -> None:
+        if span_state is None:
+            return
+        _parent_sentry_span, sentry_span = span_state
+
+        deserialized_values = json.loads(values)
+        for key, value in deserialized_values.items():
+            sentry_span.set_data(key, value)
```

Comment:
Could there be PII (Personally Identifiable Information) in that data? (Stuff like names, addresses, email addresses, phone numbers etc) If yes you should use `sentry_sdk.scope.should_send_default_pii()` to guard it.

Reply:
oh hmm, great callout

it looks like `should_send_default_pii()` (and [the Rust SDK's `send_default_pii`](https://github.com/getsentry/sentry-rust/blob/1b65b5c99af975496880e7325218479e0037d097/sentry-core/src/clientoptions.rs#L88)) are for information the integration knows is sensitive. this integration, by contrast, doesn't understand any of this data and would have to introspect or guess whether it's sensitive. so i don't think `should_send_default_pii()` is the mechanism to use here. also, this module is basically a Python port of [the Rust SDK's `tracing` integration](https://github.com/getsentry/sentry-rust/blob/master/sentry-tracing/src) and the Rust version has no particular logic around PII

as for what to do instead: is it enough to make the docs point users at the `before_send` and `before_send_transaction` hooks to provide their own PII cleanup behavior? or should this integration attempt to find obvious PII and filter it?

Reply:
I have now added some sensitive data control. So you extension uses the send_default_pii from the sentry_sdk.init() call but also can set a new paramater `send_sensitive_data` parameter to override the inits settings. 

This way the user has good control of whether they want to send sensitive data.
In case they do not want to sent sensitive data, only the values are redacted. 

See my commit: https://github.com/getsentry/sentry-python/pull/3717/commits/c66c1a04a2422c2d33895d926dbfc8c3d14b46a8